include/ibamr/MassIntegrator.h

```diff
@@ -0,0 +1,304 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2018 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_MassIntegrator
+#define included_IBAMR_MassIntegrator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include "ibamr/StaggeredStokesPhysicalBoundaryHelper.h"
+#include "ibamr/ibamr_enums.h"
+
+#include "ibtk/CartGridFunction.h"
+#include "ibtk/HierarchyGhostCellInterpolation.h"
+#include "ibtk/HierarchyMathOps.h"
+#include "ibtk/ibtk_utilities.h"
+
+#include "CoarseFineBoundary.h"
+#include "HierarchySideDataOpsReal.h"
+#include "IntVector.h"
+#include "PatchHierarchy.h"
+#include "SideVariable.h"
+#include "tbox/Database.h"
+#include "tbox/DescribedClass.h"
+#include "tbox/Pointer.h"
+
+#include <array>
+#include <string>
+#include <vector>
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM, class TYPE>
+class SAMRAIVectorReal;
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+namespace hier
+{
+template <int DIM>
+class BasePatchHierarchy;
+} // namespace hier
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class MassIntegrator is an abstract class which integrates
```

Comment:
Could this be a specialization of `HierarchyIntegrator`? There seems to be a good amount of duplicated functionality here. I would prefer to avoid introducing a new interface unless the existing one is not suitable and cannot made to be suitable.

Reply:
@boyceg `MassIntegrator` is a lightweight specialized class used to integrate the mass conservation equation at both the cell center and side center within each time step. This class does not maintain any state. It is a temporary integrator.

Reply:
I don't follow why this is not a `HierarchyIntegrator`.

Can you add some comments to the header explaining why this doesn't implement the `HierarchyIntegrator` interface?

Alternatively: It might make more sense to create a "lightweight" integrator base class that is specialized by `HierarchyIntegrator`.

Reply:
This class is integrating the mass balance PDE over a single time step and passing the mass flux to the main HieracyIntegrator.

HierarchyIntegrator's philosophy is very different. It is derived from TagAndInitializeStrategy and is responsible for data management and tagging and regridding.

OTOH, mass integrator is basically solving the PDE over one time step and giving the flux to main HierachyIntegrator (e.g., PhaseChangeHI or INSVC). It is not doing any stuff related to hierarchy management.

We can call it MassPDEIntegrator?